File: src/JobRadar.Domain/Entities/Vacancy.cs
namespace JobRadar.Domain.Entities;

/// <summary>
/// Вакансия из внешнего источника. Уникальна по паре (Source, ExternalId):
/// повторная доставка той же вакансии обновляет запись, а не создаёт дубль.
/// Идемпотентность держится на уникальном индексе + атомарном upsert
/// (VacancyUpsertService) — центральный showpiece. Системный столбец xmin
/// настроен как concurrency-token для интерактивных правок (Phase 3); путь
/// приёма его не задействует (см. JobRadarDbContext).
/// </summary>
public class Vacancy
{
    public int Id { get; set; }

    public required string Source { get; set; }
    public required string ExternalId { get; set; }

    public required string Title { get; set; }
    public string? Company { get; set; }
    public string? Market { get; set; }
    public string? Level { get; set; }
    public string? Stack { get; set; }
    public string? Location { get; set; }

    public string? SalaryRaw { get; set; }
    public long? SalaryMin { get; set; }
    public long? SalaryMax { get; set; }
    public string? SalaryCurrency { get; set; }

    public string? Skills { get; set; }
    public string? Url { get; set; }
    public DateTimeOffset? PublishedAt { get; set; }

    public DateTimeOffset FirstSeen { get; set; }
    public DateTimeOffset LastSeen { get; set; }

    // NOTE: задел под нативные вакансии работодателей (Phase «посредник»):
    // у агрегированных вакансий null, у запощенных на JobRadar — Id работодателя.
    public Guid? PostedByUserId { get; set; }
}

File: frontend/src/types/vacancy.ts
export interface VacancyDto {
  id: number;
  source: string;
  externalId: string;
  title: string;
  company: string | null;
  market: string | null;
  level: string | null;
  stack: string | null;
  location: string | null;
  salaryRaw: string | null;
  skills: string | null;
  url: string | null;
  publishedAt: string | null;
  firstSeen: string;
  lastSeen: string;
}

export interface PagedResult<T> {
  items: T[];
  page: number;
  pageSize: number;
  total: number;
  totalPages: number;
}

export type VacancyChangeOutcome = 'Inserted' | 'Updated';

export interface VacancyChangedPayload {
  source: string;
  externalId: string;
  title: string;
  company: string | null;
  market: string | null;
  level: string | null;
  stack: string | null;
  url: string | null;
  publishedAt: string | null;
  outcome: VacancyChangeOutcome;
}

export interface VacancyFilters {
  market?: string;
  level?: string;
  stack?: string;
  q?: string;
  page: number;
  pageSize: number;
}
